addons/purchase/wizard/purchase_order_cancel.py

```diff
@@ -0,0 +1,31 @@
+from odoo import _, api, models, fields
+from odoo.exceptions import UserError
+
+
+class PurchaseCancelWizard(models.TransientModel):
+    _name = 'purchase.cancel.wizard'
+    _description = 'Wizard to confirm purchase order cancellation'
+
+    purchase_order_ids = fields.Many2many('purchase.order', string="Purchase Orders")
+    purchase_count = fields.Integer(compute='_compute_purchase_count')
+
+    @api.depends('purchase_order_ids')
+    def _compute_purchase_count(self):
+        for wizard in self:
+            wizard.purchase_count = len(wizard.purchase_order_ids)
+
+    def action_confirm(self):
+        purchase_order_ids = self.purchase_order_ids.filtered(lambda p: p.state != 'cancel')
+        purchase_orders_with_receipt = purchase_order_ids.filtered(lambda po: any(line.qty_received > 0 for line in po.order_line))
+        purchase_orders_with_invoices = purchase_order_ids.filtered(lambda po: any(i.state == 'posted' for i in po.invoice_ids))
+        if purchase_orders_with_receipt and purchase_orders_with_invoices:
+            raise UserError(_('Unable to cancel selected order as either some receptions have already done or Vendor bills created'))
+        if purchase_orders_with_invoices:
+            raise UserError(_('Unable to cancel purchase order %s You must first cancel the related Vendor bills.', ', '.join(purchase_orders_with_invoices.mapped('name'))))
+        if purchase_orders_with_receipt:
+            raise UserError(_('Unable to cancel purchase order %s as some receptions have already been done.', ', '.join(purchase_orders_with_receipt.mapped('name'))))
```

Comment:
It would be nice to centralize all of this validation logic for cancelling purchase orders to the `button_cancel` method on purchase.order (we are already checking here for example if there are any linked vendor bills), and just call that method here and close the window